public/app/plugins/datasource/cloudwatch/annotationSupport.ts

```diff
@@ -0,0 +1,59 @@
+import { AnnotationQuery } from '@grafana/data';
+import { AnnotationQueryEditor } from './components/AnnotationQueryEditor';
+import { isCloudWatchAnnotation } from './guards';
+import { CloudWatchAnnotationQuery, CloudWatchQuery, LegacyAnnotationQuery } from './types';
+
+export const CloudWatchAnnotationSupport = {
+  // converts legacy angular style queries to new format. Also sets the same default values as in the deprecated angular directive
+  prepareAnnotation: (
+    query: LegacyAnnotationQuery | AnnotationQuery<CloudWatchAnnotationQuery>
+  ): AnnotationQuery<CloudWatchAnnotationQuery> => {
+    if (isCloudWatchAnnotation(query)) {
+      return query;
+    }
+
+    return {
+      // setting AnnotationQuery props explicitly since spreading would incorrectly use props that should be on the target only
+      datasource: query.datasource,
+      enable: query.enable,
+      iconColor: query.iconColor,
+      name: query.name,
+      builtIn: query.builtIn,
+      hide: query.hide,
+      target: {
+        ...query.target,
+        ...query,
+        statistic: query.statistic || 'Average',
+        region: query.region || 'default',
+        queryMode: 'Annotations',
+        refId: query.refId || 'annotationQuery',
+      },
+    };
+  },
+  // return undefined if query is not complete so that annotation query execution is quietly skipped
+  prepareQuery: (anno: AnnotationQuery<CloudWatchAnnotationQuery>): CloudWatchQuery | undefined => {
+    if (!anno.target) {
+      return undefined;
+    }
+
+    const {
+      prefixMatching,
+      actionPrefix,
+      alarmNamePrefix,
+      statistic,
+      namespace,
+      metricName,
+      dimensions = {},
+    } = anno.target;
+    const validPrefixMatchingQuery = !!prefixMatching && !!actionPrefix && !!alarmNamePrefix;
+    const validMetricStatQuery =
+      !prefixMatching && !!namespace && !!metricName && !!statistic && !!Object.values(dimensions).length;
+
+    if (validPrefixMatchingQuery || validMetricStatQuery) {
+      return anno.target;
+    }
+
+    return undefined;
```

Comment:
this is a bit risky, it could hide legit errors, for example when migrating an unexpected legacy query, but the other option is to annoyingly show errors to users while writing these annotations so I guess it's fine.

Reply:
not sure I follow - legacy queries should be migrated in `prepareAnnotation` ?

Reply:
yes, I mean, if `prepareAnnotation` had a bug and were returning an incomplete target, this would silently do nothing rather than returning an error but it's probably fine.

Reply:
It's the same as in normal query executions path - incomplete queries are filtered out and silently skipped.